lib/config/param.c

```diff
@@ -314,19 +317,24 @@ param_default_converter(const struct param_spec *ps, const cJSON *node, void *va
 
             to_conv = cJSON_GetNumberValue(node);
             if (!IS_WHOLE(to_conv)) {
-                log_err("%s must be a whole number", ps->ps_name);
+                log_err("Value of %s must be a whole number", ps->ps_name);
                 return false;
             }
-            if (to_conv < INT64_MIN || to_conv > (double)INT64_MAX) {
+
+            /* Converting a double to an int64_t is extremely difficult.
+             * Confirm that the double is a whole number, convert it to a
+             * string, and parse the string to an int64_t.
+             */
+            snprintf(buf, sizeof(buf), "%.0f", to_conv);
+            tmp = strtoll(buf, &end, 10);
+            if (*end != '\0') {
                 log_err(
-                    "Value of %s must be greater than or equal to %ld and less than or equal to "
-                    "%ld",
+                    "Value of %s must be greater than or equal to 0 and less than or equal to %ld",
                     ps->ps_name,
-                    INT64_MIN,
                     INT64_MAX);
                 return false;
             }
-            *(int64_t *)value = (int64_t)to_conv;
+            *(int64_t *)value = tmp;
```

Comment:
Hmm, not sure what the point of this is..  You should send the param as a string not as a double, then decode it here with stroul() or whatever..

Reply:
If I've already deserialized the JSON, this is impossible.

Reply:
Huh?  Why are you putting it in a double in the first place?  What am I missing?  If you encode it as a string don't you get back a string when you deserialize it?

Reply:
I am putting it into a double because that is the cJSON API. I have no control over what the return type of `cJSON_GetNumberValue()` is.

I don't understand your question.

Reply:
Putting it into a double causes all this insane code to extract it, which is lossy should the number require more than 52 bits to represent.  Why not just put it in as a string?  This code you're written to extract doesn't solve any problems as far as I can tell, it cannot detect that any precision was lost, so what's the point?

Reply:
Are you saying we should have no integers in our configuration formats?

> This code you're written to extract doesn't solve any problems as far as I can tell, it cannot detect that any precision was lost, so what's the point?

But it does solve a problem, the undefined behavior that clang was warning us about.

Reply:
Information is lost during the encoding, and the receiver cannot even discern that information was lost.  So why go down this flawed path?  It's only going to end in misery at some point...
Where do you get any string?  With cJSON you have to snprintf into a buffer, strdup it, and hand it to cJSON.
clang is warning that you're doing something unsavory, AFAICT this new code is just trying to paper over the problem, while the solution that will work for all time is to fix it at the sender.

Reply:
This code path is hit when a config file is parsed and when a `key=value` string is passed. I have no desire to hand-parse JSON on the right hand side of the equal sign.

> With cJSON you have to snprintf into a buffer, strdup it, and hand it to cJSON.

What buffer and strdup are you referring to?

Regardless, I am not going to change this code path right now. The only actionable thing I see is picking a better JSON library. This one is not that great.

Reply:
The fact that cJSON forces all numeric values through a double is reason alone to ditch cJSON.

We can't solve the precision loss problem right now unless cJSON lets you get a numeric value in raw string form (before cJSON parsed it as a double).

Do we have any params that need 64-bit values large enough to run into problems here?